### PR TITLE
update AssemblyScript to 0.9.1 + minor imrovements

### PR DESCRIPTION
**Pull request by MaxGraey**

### CODE REVIEW

File: src/assembly/gx_texture_as.ts [hunk 1]
```diff
@@ -49,13 +49,15 @@ export function decode_I4(pScratch: u32, pDst: u32, pSrc: u32, w: u32, h: u32):
     for (let yy: u32 = 0; yy < h; yy += 8) {
         for (let xx: u32 = 0; xx < w; xx += 8) {
             for (let y = 0; y < 8; y++) {
+                let yyy = yy + y;
+                if (yyy >= h) continue;
                 for (let x = 0; x < 8; x++, srcOffs++) {
-                    if (xx + x >= w || yy + y >= h)
```
Comment (magcius) [inline src/assembly/gx_texture_as.ts]: Can the compiler not do this CSE?

File: src/assembly/gx_texture_as.ts [hunk 2]
```diff
@@ -49,13 +49,15 @@ export function decode_I4(pScratch: u32, pDst: u32, pSrc: u32, w: u32, h: u32):
...
-                        continue;
-                    let dstPixel: u32 = (w * (yy + y)) + xx + x;
+                    let xxx = xx + x;
+                    if (xxx >= w) continue;
+                    let dstPixel: u32 = w * yyy + xxx;
                     let dstOffs = pDst + dstPixel * 4;
                     let ii: u8 = get(pSrc + (srcOffs >>> 1));
-                    let i4: u8 = ii >>> ((srcOffs & 1) ? 0 : 4) & 0x0F;
+                    let i4: u8 = ii >>> ((srcOffs & 1) << 2 as u8) & 0x0F;
```
Comment (magcius) [inline src/assembly/gx_texture_as.ts]: Please parenthesize to disambiguate precedence between `<<` and `as`. I tried to look in the TypeScript spec to find the correct placement of `as` in the binop tree, but it's [not yet documented](https://github.com/microsoft/TypeScript/blob/master/doc/spec.md#416-type-assertions).

File: src/assembly/gx_texture_as.ts (hunk 1 above)
Comment (MaxGraey) [reply src/assembly/gx_texture_as.ts]: Yes, CSE possible here but not LICM (loop invariant motion).

**Comment by magcius** [comment]
FYI, I've reverted this as it had regressions. Simplest test case is Smash Melee.

https://noclip.website/#SuperSmashBrosMelee/Title;ShareData=AQJgOUhU@=8PHu;87LG799xE25&pyFUbg@e9Z;8Z+fWDNTw,rWUAv(_UVVIzVS

**Comment by MaxGraey** [comment]
Hmm, looks like the last commit was superfluous.

**Pull request merged**